blscurve/bls_batch_verifier.nim

```diff
@@ -0,0 +1,372 @@
+# Nim-BLSCurve
+# Copyright (c) 2018-Present Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+import
+  ./bls_backend, ./bls_sig_min_pubkey,
+  ./openmp
+
+# BLS Batch Verifier
+# ----------------------------------------------------------------------
+# We use OpenMP here but might want to use a simple threadpool
+# for portability as Mac compiler doesn't ship with OpenMP
+# and while MSVC has OpenMP, it's unsure about Mingw
+#
+# Also as Nim supports for view types and openarray in object fields is experimental
+# we collect/copy the inputs in the object for now by copy.
+# instead of accumulating them in a pairing context.
+# Note that a pairing context is 3+MB
+# while we have
+# - publickey = 48B
+# - signature = 96B
+# - message = 32B (assuming sha256 hashed)
+# hence 176B only
+#
+# TODO: Once ContextMultiAggregateVerify is implemented
+# for Milagro/Miracl, this wouldn't need to be in the BLST specific file
+
+type
+  SignatureSet = object
+    ## A (Public Key, Message, Signature) triplet
+    ## that will be batch verified.
+    ## This should not hold GC-ed memory
+    ## as this would complexify multithreading.
+    ## Consequently this assumes that message
+    ## is the output of a fixed size hash function.
+    signature: Signature
+    pubkey: PublicKey
+    message: array[32, byte]
+
+  BatchedBLSVerifier* = object
+    ## A type to batch BLS multi signatures (aggregated or individual)
+    ## verification using multiple cores if compiled with OpenMP
+    sets: seq[SignatureSet]
+
+    # Per-batch contexts for multithreaded batch verification
+    batchContexts: seq[ContextMultiAggregateVerify[DST]]
+    updateResults: seq[tuple[ok: bool, padCacheLine: array[64, byte]]]
+
+func init*(T: type BatchedBLSVerifier): T {.inline.} =
+  ## Initialize or reinitialize a batchedBLS Verifier
+  discard # default initialization
+
+func clear*(batcher: var BatchedBLSVerifier) {.inline.} =
+  ## Initialize or reinitialize a batchedBLS Verifier
+  batcher.sets.setLen(0)
+  batcher.batchContexts.setLen(0)
+
+func incl*(
```

Comment:
this should be `add` since there's neither a constraint nor a check that the same signature is checked twice - ie it's not a set operation